artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ObjectPool.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.utils;
+
+import java.util.Queue;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import io.netty.util.internal.PlatformDependent;
+
+/**
+ * A simple encapsulation of Netty MpscQueue to provide a pool of objects.
+ * @param <T>
+ */
+public class ObjectPool<T> {
+
+   private final Queue<T> internalPool;
+
+   private final Consumer<T> cleaner;
+   private final Supplier<T> supplier;
+
+   public ObjectPool(int maxSize, Consumer<T> cleaner, Supplier<T> supplier) {
+      if (maxSize > 0) {
+         internalPool = PlatformDependent.newFixedMpscQueue(maxSize);
```

Comment:
This is the complex part: this queue work correctly only if there are multiple producers and a single consumer ie it breaks with many parallel consumers.
I am not sure I would create a general purpose utility for this reason, because the use case is very narrow and difficult to be enforced...
In addition, it seems that's used in case no pool is needed, that means 

1) we pay several additional costs just to create the other object
2) we always create the object pool itself

We can use a JMH bench so I can show better what I mean

Reply:
In short:
- If you still want to create the utility I would use different factory method and config parameters to cover all the use cases (spsc, mpmc,mpsc,spmc) but probably there are not enouf JCTools queues and just using mpmc is not a good idea because the performance and memory footprint are not negligible
- I wouldn't create any object pool in case is not needed, to save additional methods in the stack, lambdas and creation checks

Reply:
In addition: I see that probably I've been too conservative and mpsc is too much, just using spsc should be enough, please take a look where we're going to release the used packet instances -> it should always happen in a "as single-thread fashion" ie not the same thread, but there are no concurrent access from more then 1 thread each time to release it...but I could be wrong :D

Reply:
@franz1981 this PR is pretty much doing the same thing you did.. I'm just encapsulating MPSC into an object.

This is just to be clearer to read.

When I asked you to review the code here was to review the API.

Now if we should pool or not it's an orthogonal discussion. The code right now is pooling and I would like it (tried it) to be more readable on its intention.

Reply:
I would still create the Utility. document its usage...

I would rather have the pool.borrow in the code with the logic encapsulated. It is prettier.

Reply:
@franz1981 I renamed the ObjectPool class, and documented the usage.